pyanaconda/kickstart.py

```diff
@@ -386,26 +409,39 @@ def execute(self, storage, ksdata, instClass):
                             formatErrorMsg(self.lineno,
                                            msg=(_("Requested boot drive \"%s\" doesn't exist or cannot be used.")
                                                 % self.bootDrive)))
+        # Take valid disk from --driveorder
         elif len(self.driveorder) >= 1:
-            log.debug("Bootloader: use '%s' first disk from driveorder as boot drive",
-                      self.driveorder[0])
+            log.debug("Bootloader: use '%s' first disk from driveorder as boot drive, dry run %s",
+                      self.driveorder[0], dry_run)
             self.bootDrive = self.driveorder[0]
         else:
-            boot_drive = None
             # Try to find /boot
-            for part in ksdata.partition.partitions:
-                if part.mountpoint == "/boot":
-                    device_match = deviceMatches(part.disk, devicetree=storage.devicetree)
-                    if len(device_match) == 1 and device_match[0] in disk_names:
-                        log.debug("Bootloader: use /boot partition '%s' as boot drive", device_match[0])
-                        boot_drive = device_match[0]
-                    break
-            else: # Nothing was found use first disk from Blivet
-                log.debug("Bootloader: fallback use first disk return from Blivet '%s' as boot drive",
-                          disk_names[0])
-                boot_drive = disk_names[0]
-
-            self.bootDrive = boot_drive
+            #
+            # This method is executed two times. Before and after partitioning.
+            # In the first run, the result is used for other partitioning but
+            # the second will be used.
+            try:
+                boot_dev = storage.mountpoints["/boot"]
+            except KeyError:
+                log.debug("Bootloader: /boot partition is not present, dry run %s", dry_run)
+            else:
+                boot_drive = ""
+                # Find disk ancestor
+                for ancestor in boot_dev.ancestors:
```

Comment:
They have the advantage of being, by definition, disks, which would simplify your code a bit.